src/Draco.Compiler/Syntax/ParseTreeVisitor.cs

```diff
@@ -0,0 +1,147 @@
+    public abstract T VisitLabelDecl(Decl.Label decl);
+    public abstract T VisitVariableDecl(Decl.Variable decl);
+
+    public abstract T VisitFuncParam(FuncParam param);
+
+    public virtual T VisitFuncBody(FuncBody body) => body switch
+    {
+        FuncBody.BlockBody block => this.VisitBlockFuncBody(block),
+        FuncBody.InlineBody inline => this.VisitInlineFuncBody(inline),
+
+        _ => throw new InvalidOperationException()
+    };
+    public abstract T VisitBlockFuncBody(FuncBody.BlockBody body);
+    public abstract T VisitInlineFuncBody(FuncBody.InlineBody body);
+
+    public virtual T VisitTypeExpr(TypeExpr typeExpr) => typeExpr switch
+    {
+        TypeExpr.Name name => this.VisitNameTypeExpr(name),
+
+        _ => throw new InvalidOperationException()
+    };
+    public abstract T VisitNameTypeExpr(TypeExpr.Name typeExpr);
+
+    public abstract T VisitTypeSpecifier(TypeSpecifier specifier);
+
+    public virtual T VisitStmt(Stmt stmt) => stmt switch
+    {
+        Stmt.Decl decl => this.VisitDeclStmt(decl),
+        Stmt.Expr expr => this.VisitExprStmt(expr),
+
+        _ => throw new InvalidOperationException()
+    };
+    public abstract T VisitDeclStmt(Stmt.Decl stmt);
+    public abstract T VisitExprStmt(Stmt.Expr stmt);
+
+    public virtual T VisitExpr(Expr expr) => expr switch
+    {
+        Expr.Block block => this.VisitBlockExpr(block),
+        Expr.If @if => this.VisitIfExpr(@if),
+        Expr.While @while => this.VisitWhileExpr(@while),
+        Expr.Goto @goto => this.VisitGotoExpr(@goto),
+        Expr.Return @return => this.VisitReturnExpr(@return),
+        Expr.Literal literal => this.VisitLiteralExpr(literal),
+        Expr.FuncCall funcCall => this.VisitFuncCallExpr(funcCall),
+        Expr.Index index => this.VisitIndexExpr(index),
+        Expr.Variable variable => this.VisitVariableExpr(variable),
+        Expr.MemberAccess memberAccess => this.VisitMemberAccessExpr(memberAccess),
+        Expr.Unary unary => this.VisitUnaryExpr(unary),
+        Expr.Binary binary => this.VisitBinaryExpr(binary),
+        Expr.Grouping grouping => this.VisitGroupingExpr(grouping),
+
+        _ => throw new InvalidOperationException()
+    };
+    public abstract T VisitBlockExpr(Expr.Block expr);
+    public abstract T VisitIfExpr(Expr.If expr);
+    public abstract T VisitWhileExpr(Expr.While expr);
+    public abstract T VisitGotoExpr(Expr.Goto expr);
+    public abstract T VisitReturnExpr(Expr.Return expr);
+    public abstract T VisitLiteralExpr(Expr.Literal expr);
+    public abstract T VisitFuncCallExpr(Expr.FuncCall expr);
+    public abstract T VisitIndexExpr(Expr.Index expr);
+    public abstract T VisitVariableExpr(Expr.Variable expr);
+    public abstract T VisitMemberAccessExpr(Expr.MemberAccess expr);
+    public abstract T VisitUnaryExpr(Expr.Unary expr);
+    public abstract T VisitBinaryExpr(Expr.Binary expr);
+    public abstract T VisitGroupingExpr(Expr.Grouping expr);
```

Comment:
Why do these don't do their recursive thing?

Reply:
Sure these could visit all of their subexpressions, but what value should they return? `default(T)`?

Reply:
Also I'm assuming you'd have to call `base.VisitWhatever(whatever)` if you want to override any of these?

Reply:
`default(T)` or a virtual default providing prop are both fine. And yes, you need to call base to recurse, just like ANTLR visitors